### PR TITLE
CVSL-2157 Moving date manipulation to the boundaries of the application

### PR DESCRIPTION
**Pull request by andrewrlee**

### CODE REVIEW

File: server/@types/assessForEarlyReleaseApiClientTypes.ts [hunk 1]
```diff
@@ -1,5 +1,14 @@
...
 
-export type OffenderSummary = components['schemas']['OffenderSummary']
-export type AssessmentSummary = components['schemas']['AssessmentSummary']
-export type InitialChecks = components['schemas']['InitialChecks']
+type ParsingDates<T, Fields> = {
+  [K in keyof T]: K extends Fields ? Date : T[K] extends object ? ParsingDates<T[K], Fields> : T[K]
+}
+
+export type _OffenderSummary = components['schemas']['OffenderSummary']
+export type OffenderSummary = ParsingDates<_OffenderSummary, 'hdced'>
```
Comment (andrewrlee) [inline server/@types/assessForEarlyReleaseApiClientTypes.ts:8]: this basically rewrites the type to have certain fields as dates. 

It seems to work recursively to - so initial checks has `hdced` on a nested object and that also gets rewritten. 

Obviously the field have to be manually converted, see below..

File: server/data/assessForEarlyReleaseApiClient.ts [hunk 2]
```diff
@@ -15,29 +22,36 @@ export default class AssessForEarlyReleaseApiClient {
...
   async getCaseAdminCaseload(prisonCode: string): Promise<OffenderSummary[]> {
-    const caseAdminCaseload = await this.restClient.get<OffenderSummary[]>({
+    const caseAdminCaseload = await this.restClient.get<_OffenderSummary[]>({
       path: `/prison/${prisonCode}/case-admin/caseload`,
     })
     return caseAdminCaseload.map(c => {
-      return { ...c, hdced: format(c.hdced, 'dd MMM yyyy') }
+      return { ...c, hdced: parseIsoDate(c.hdced) }
     })
   }
```
Comment (andrewrlee) [inline server/data/assessForEarlyReleaseApiClient.ts:31]: this receives `_OffenderSummary` from the API where `hdced` is a string.
It returns a `OffenderSummary` where `hdced` is a date.. this only compiles as we specifically parse the date on line 29.

File: server/routes/caseAdmin/caseload.ts [hunk 3]
```diff
@@ -14,15 +13,13 @@ export default class CaseloadRoutes {
 
     const caseload = offenderSummaryList.map(offender => {
       return {
-        name: convertToTitleCase(`${offender.forename} ${offender.surname}`.trim()),
-        createLink: this.findCreateLinkToDisplay(offender.prisonNumber),
+        createLink: paths.prison.assessment.home(offender),
```
Comment (andrewrlee) [inline server/routes/caseAdmin/caseload.ts:16]: moved to use static path, and pushed other logic down into the client

File: server/views/pages/caseAdmin/caseload.njk [hunk 4]
```diff
@@ -31,9 +31,9 @@
         {
             attributes: {
                 id: 'hdced-' + loop.index,
-                "data-sort-value": offender.hdced
+                "data-sort-value": offender.hdced | toMillis
             },
-            html: offender.hdced
+            html: offender.hdced | formatDate('dd MMM yyyy')
```
Comment (andrewrlee) [inline server/views/pages/caseAdmin/caseload.njk:36]: now only do formatting of date in templates

**Review by justin-gardiner — APPROVED** [review]
LGTM

**Review by justin-gardiner — DISMISSED** [review]
LGTM

**Review by justin-gardiner — APPROVED** [review]
LGTM

**Pull request merged**